DataScience/ProbabilisticParser/data/createTrainingDataProbabilisticParser.py

```diff
@@ -0,0 +1,290 @@
+    :type path: str
+    :param filename: name of the address base PAF file
+    :type filename: str
+    :param training_sample_size: number of training samples, if exceeds the number of examples then no holdout data
+    :type training_sample_size: int
+    :param holdout_sample_size: number of holdout samples, if exceeds the number of potential holdouts, then use all
+    :type holdout_sample_size: int
+    :param training_subsamples: a tuple of containing the number of samples to use for subsamples drawn
+                                from the training data
+    :type training_subsamples: tuple
+    :param out_path: location where to store the output files
+    :type out_path: str
+    :param outfile: name of the training data file
+    :type outfile: str
+    :param holdout_file: name of the holdout data file
+    :type holdout_file: str
+
+    :return: None
+    """
+    # delivery point information used for training a probabilistic parser
+    columns = {'ORGANISATION_NAME': 'OrganisationName',
+               'DEPARTMENT_NAME': 'DepartmentName',
+               'SUB_BUILDING_NAME': 'SubBuildingName',
+               'BUILDING_NAME': 'BuildingName',
+               'BUILDING_NUMBER': 'BuildingNumber',
+               'THROUGHFARE': 'StreetName',
+               'DEPENDENT_LOCALITY': 'Locality',
+               'POST_TOWN': 'TownName',
+               'POSTCODE': 'Postcode',
+               'WELSH_THOROUGHFARE': 'WelshStreetName',
+               'WELSH_DEPENDENT_LOCALITY': 'WelshLocality',
+               'WELSH_POST_TOWN': 'WelshTownName',
+               'PO_BOX_NUMBER': 'PObox'}
+
+    data = pd.read_csv(path + filename, dtype=str, usecols=columns.keys())
+
+    print('Data source contains', len(data.index), 'addresses')
+    print('removing STREET RECORDs, PARKING SPACEs, PONDs etc...')
+    # remove street records from the list of potential matches - this makes the search space slightly smaller
+    exclude = 'STREET RECORD|ELECTRICITY SUB STATION|PUMPING STATION|POND \d+M FROM|PUBLIC TELEPHONE|'
+    exclude += 'PART OF OS PARCEL|DEMOLISHED BUILDING|CCTV CAMERA|TANK \d+M FROM|SHELTER \d+M FROM|TENNIS COURTS|'
+    exclude += 'PONDS \d+M FROM|SUB STATION|PARKING SPACE'
+    msk = data['BUILDING_NAME'].str.contains(exclude, na=False, case=False) | \
+          data['THROUGHFARE'].str.contains(exclude, na=False, case=False)
+    data = data.loc[~msk]
+    print('After removing parking spaces etc.', len(data.index), 'addresses remain')
+
+    # remove PO BOXes
+    # todo: confirm what to do with PO BOXes - separate process of part of the parser? Latter -> new label
+    msk = data['PO_BOX_NUMBER'].isnull()
+    data = data.loc[msk]
+    print('After removing PO boxes', len(data.index), 'addresses remain')
+
+    print('Replacing English spelling with Welsh names for those addresses where Welsh is available')
+    msk = data['WELSH_THOROUGHFARE'].notnull() & data['WELSH_DEPENDENT_LOCALITY'].notnull() & \
```

Comment:
Python distinguishes between & and "and" operators, the latter being boolean and the first one being bitwise. It goes back to whether one is using pure Python or NumPy powered, see the example here:
http://stackoverflow.com/questions/22646463/difference-between-and-boolean-vs-bitwise-in-python-why-difference-i
Not ideal if you ask me...